artifactory/commands/python/twine.go

```diff
@@ -0,0 +1,196 @@
+}
+
+func (tc *TwineCommand) SetServerDetails(serverDetails *config.ServerDetails) *TwineCommand {
+	tc.serverDetails = serverDetails
+	return tc
+}
+
+func (tc *TwineCommand) SetTargetRepo(targetRepo string) *TwineCommand {
+	tc.targetRepo = targetRepo
+	return tc
+}
+
+func (tc *TwineCommand) SetArgs(args []string) *TwineCommand {
+	tc.args = args
+	return tc
+}
+
+func (tc *TwineCommand) Run() (err error) {
+	// Assert no forbidden flags were provided.
+	if tc.isRepoConfigFlagProvided() {
+		return errorutils.CheckErrorf(tc.getRepoConfigFlagProvidedErr())
+	}
+	if err = tc.extractAndFilterArgs(tc.args); err != nil {
+		return err
+	}
+	callbackFunc, err := tc.setAuthEnvVars()
+	defer func() {
+		err = errors.Join(err, callbackFunc())
+	}()
+
+	collectBuild, err := tc.buildConfiguration.IsCollectBuildInfo()
+	if err != nil {
+		return err
+	}
+	// If build info is not collected, or this is not an upload command, run the twine command directly.
+	if !collectBuild || tc.commandName != _uploadCmdName {
+		return tc.runPlainTwineCommand()
+	}
+	return tc.uploadAndCollectBuildInfo()
+}
+
+func (tc *TwineCommand) extractAndFilterArgs(args []string) (err error) {
+	cleanArgs := append([]string(nil), args...)
+	cleanArgs, tc.buildConfiguration, err = buildUtils.ExtractBuildDetailsFromArgs(cleanArgs)
+	if err != nil {
+		return
+	}
+	tc.args = cleanArgs
+	return
+}
+
+func (tc *TwineCommand) setAuthEnvVars() (callbackFunc func() error, err error) {
+	oldRepoUrl := os.Getenv(_repositoryUrlEnvKey)
+	oldUsername := os.Getenv(_usernameEnvKey)
+	oldPassword := os.Getenv(_passwordEnvKey)
+	callbackFunc = func() error {
+		return errors.Join(os.Setenv(_repositoryUrlOptionKey, oldRepoUrl), os.Setenv(_usernameEnvKey, oldUsername), os.Setenv(_passwordEnvKey, oldPassword))
+	}
+
+	if err = os.Setenv(_repositoryUrlEnvKey, utils.AddTrailingSlashIfNeeded(tc.serverDetails.ArtifactoryUrl)+_apiPypi+tc.targetRepo); err != nil {
+		return
+	}
+
+	username := tc.serverDetails.User
+	password := tc.serverDetails.Password
+	// Get credentials from access-token if exists.
+	if tc.serverDetails.GetAccessToken() != "" {
+		if username == "" {
+			username = auth.ExtractUsernameFromAccessToken(tc.serverDetails.GetAccessToken())
+		}
+		password = tc.serverDetails.GetAccessToken()
+	}
+
+	if err = os.Setenv(_usernameEnvKey, username); err != nil {
+		return
+	}
+	err = os.Setenv(_passwordEnvKey, password)
+	return
+}
+
+func (tc *TwineCommand) runPlainTwineCommand() error {
+	log.Debug("Running twine command:", tc.commandName, strings.Join(tc.args, " "))
+	args := append([]string{tc.commandName}, tc.args...)
+	cmd := exec.Command(_twineExecName, args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+	return err
+}
+
+func (tc *TwineCommand) uploadAndCollectBuildInfo() error {
+	buildInfo, err := buildUtils.PrepareBuildPrerequisites(tc.buildConfiguration)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if buildInfo != nil && err != nil {
```

Comment:
What's the reason we're cheching here whether err isn't nil?
Isn't it enough to check that buildInfo isn't nil?